service/worker/batcher/activities.go

```diff
@@ -551,22 +590,57 @@ func startTaskProcessorProtobuf(
 			default:
 				err = errors.New(fmt.Sprintf("unknown batch type: %v", batchOperation.BatchType))
 			}
-			if err != nil {
-				metrics.BatcherProcessorFailures.With(metricsHandler).Record(1)
-				logger.Error("Failed to process batch operation task", tag.Error(err))
-				nonRetryable := slices.Contains(batchOperation.NonRetryableErrors, err.Error())
-				if nonRetryable || task.attempts > int(batchOperation.AttemptsOnRetryableError) {
-					respCh <- taskResponse{err: err, page: task.page}
-				} else {
-					// put back to the channel if less than attemptsOnError
-					task.attempts++
-					taskCh <- task
-				}
-			} else {
-				metrics.BatcherProcessorSuccess.With(metricsHandler).Record(1)
-				respCh <- taskResponse{err: nil, page: task.page}
-			}
+			a.handleTaskResult(batchOperation, task, err, taskCh, respCh, metricsHandler, logger)
+		}
+	}
+}
+
+func (a *activities) handleTaskResult(
+	batchOperation *batchspb.BatchOperationInput,
+	task task,
+	err error,
+	taskCh chan task,
+	respCh chan taskResponse,
+	metricsHandler metrics.Handler,
+	logger log.Logger,
+) {
+	if err != nil {
+		metrics.BatcherProcessorFailures.With(metricsHandler).Record(1)
+		logger.Error("Failed to process batch operation task", tag.Error(err))
+		nonRetryable := slices.Contains(batchOperation.NonRetryableErrors, err.Error())
+		if nonRetryable || task.attempts > int(batchOperation.AttemptsOnRetryableError) {
+			respCh <- taskResponse{err: err, page: task.page}
+		} else {
+			// put back to the channel if less than attemptsOnError
+			task.attempts++
+			taskCh <- task
 		}
+	} else {
+		metrics.BatcherProcessorSuccess.With(metricsHandler).Record(1)
+		respCh <- taskResponse{err: nil, page: task.page}
+	}
+}
+
+func (a *activities) processAdminTask(
+	ctx context.Context,
+	batchOperation *batchspb.BatchOperationInput,
+	task task,
+	limiter *rate.Limiter,
+) error {
+	adminReq := batchOperation.AdminRequest
+	switch op := adminReq.Operation.(type) {
+	case *adminservice.StartAdminBatchOperationRequest_RefreshWorkflowTasksOperation:
+		return processTask(ctx, limiter, task,
+			func(execution *commonpb.WorkflowExecution) error {
+				_, err := a.AdminClient.RefreshWorkflowTasks(ctx, &adminservice.RefreshWorkflowTasksRequest{
+					NamespaceId: batchOperation.NamespaceId,
+					Execution:   execution,
+					Archetype:   op.RefreshWorkflowTasksOperation.GetArchetype(),
```

Comment:
let's retrieve this from visibility records itself. You can follow what I did in this PR: https://github.com/temporalio/temporal/pull/8768/changes#diff-dc576c64d23020364cc358402a580e400ec6f3c18015af377e2540db66fbaafaR186.

I am also ok with calling history client refresh tasks api directly (so that you don't have to convert archetypeID to archetype name) as well as long as the outbound call is captured by OPS rate limiter.

Changing the data structure used in the activity's pagination logic shouldn't cause any compatibility issue as everything is in memory.

Reply:
Speaking of OPS rate limiter, let's use CallerTypePreemptable when making the call. Probably set it at the beginning of the activity when this is a admin batch op.